libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1379,14 +1379,13 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		{
 			solUnimplementedAssert(!type.isDynamicallyEncoded(), "");
 			m_context << type.calldataOffsetOfMember(member) << Instruction::ADD;
+			// For non-value types the calldata offset is returned directly.
 			if (_memberAccess.annotation().type->isValueType())
 			{
-				CompilerUtils(m_context).loadFromMemoryDynamic(
-					*_memberAccess.annotation().type,
-					true,
-					true,
-					false
-				);
+				m_context << Instruction::DUP1;
+				m_context << type.calldataEncodedSize(true);
+				m_context << Instruction::ADD;
+				CompilerUtils(m_context).abiDecode({_memberAccess.annotation().type}, false);
```

Comment:
For uint256, this will turn into an assembly function call where the function only consists of a single `calldataload`, which cannot be inlined by the current optimizer. This is quite inefficient. Should we make a small exception for types where storageBytes is 32 and call `loadFromMemoryDynamic` as above? Actually `loadFromMemoryDynamic` also performs range checks. So another solution would be to update `CompilerUtils::convertType(` together with https://github.com/ethereum/solidity/pull/5815

Reply:
storageBytes is probably right, but feels a bit off... ``calldataEncodedSize(false)`` should do it as well, shouldn't it? And I'm wondering whether this actually needs to range-check that we're inside calldatasize or whether we can just use ``calldataload`` directly if ``calldataEncodedSize(false)==32``... probably enough for the statically encoded case, right?

Reply:
We do not need any calldatasize-range checks here for any value type. Using calldataEncodedSize is better, yes. And you can just use `calldataload`, yes :)

Reply:
Or... we could just always call ``CompilerUtils(m_context).loadFromMemoryDynamic(*_memberAccess.annotation().type, true, false, false)`` here, that should do it and degenerate to a single ``calldataload`` whenever possible - I guess that's what you meant with updating ``CompilerUtils::convertType(`` together with #5815, so that (at least in this case) ``loadFromMemoryDynamic`` will validate instead of clean?

Reply:
Yes, exactly.